### PR TITLE
added type hints for  __contains__

### LINKED ISSUE #129646: [typing] Annotations missing for `Tensor.__contains__`
### 🐛 Describe the bug

```python
from torch import Tensor
reveal_type(Tensor.__contains__)
# mypy:  "def (torch._tensor.Tensor, Any) -> Any"
# pyright: "(self: Tensor, element: Unknown) -> (Any | int | float | bool | Unknown)"
```

Both type checkers infer the signature from https://github.com/pytorch/pytorch/blob/ff026f3d0a68a4533b193498adc5b5c7b045b73c/torch/_tensor.py#L1106-L1107

Because annotations for `__contains__` are missing in `torch/_C/__init__.pyi.in`

### Versions

[pip3] mypy==1.10.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==2.0.0
[pip3] torch==2.3.1
[pip3] torchinfo==1.8.0

cc @ezyang @malfet @xuzhao9 @gramster

### PR DESCRIPTION
**Pull request by randolf-scholz**
- Fixes #129646
- Added test in test/typing/reveal/tensor_constructors.py

cc @ezyang @malfet @xuzhao9 @gramster

### CODE REVIEW

File: tools/pyi/gen_pyi.py [hunk 1]
```diff
@@ -1101,6 +1101,7 @@ def replace_special_case(hint: str) -> str:
                     )
                 )
             ],
+            "__contains__": ["def __contains__(self, other: Any) -> _bool: ..."],
```
Comment (randolf-scholz) [inline tools/pyi/gen_pyi.py:1104]: This is doubled, considering that the type hint in [torch/_tensor.py](https://github.com/pytorch/pytorch/pull/129653/files#diff-41eb33210e32f7a565a420a78573740a643b5f124726e7013e8b96aad158948e) should take priority, but some type checkers might ignore it when they detect the pyi file.

**Comment by colesbury** [comment]
@Skylion007, would you please review this PR?

File: torch/_tensor.py [hunk 2]
```diff
@@ -1116,7 +1116,7 @@ def __contains__(self, element):
             element, (torch.Tensor, Number, torch.SymInt, torch.SymFloat, torch.SymBool)
         ):
             # type hint doesn't understand the __contains__ result array
-            return (element == self).any().item()  # type: ignore[union-attr]
+            return bool((element == self).any().item())  # type: ignore[union-attr]
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_tensor.py:1119]: type ignore is probably safer, although I guess it looks like `__contains__` gets coerced to bool anyway:

```
>>> class F:
...   def __contains__(self, x):
...     return 2
... 
>>> 2 in F()
True
```

**Review by ezyang — APPROVED** [review]

**Comment by randolf-scholz** [comment]
The previous merge failed due to some apparantly unrelated CI issue.
I also pushed a minor edit converting to posititional-only arguments (So that e.g. a Protocol class that uses a positional_or_keyword parameter instead still matches.)

**Pull request closed without merging**